clients/src/main/java/org/apache/kafka/common/protocol/types/Schema.java

```diff
@@ -63,15 +63,32 @@ public void write(ByteBuffer buffer, Object o) {
         }
     }
 
-    /**
-     * Read a struct from the buffer
-     */
     @Override
     public Struct read(ByteBuffer buffer) {
+        return read(buffer, false);
+    }
+
+    /**
+     * Read a struct from the buffer. Missing fields at the end of the buffer are replaced with
+     * their default values if such fields are optional; otherwise a {@code SchemaException} is
+     * thrown to signify that mandatory fields are missing.
+     */
+    public Struct read(ByteBuffer buffer, boolean tolerateMissingWithDefaults) {
```

Comment:
As an alternative, we could let this flag be part of the schema definition. In other words, the creator of the schema must decide whether to allow optional fields at the end of the schema. Did you have any reason to prefer leaving the decision to the parser?

Reply:
No particular reason. Added it in the constructor now. Has to go before the var-arg.